mlflow/langchain/_langchain_autolog.py

```diff
@@ -33,6 +38,141 @@
+            log_model_signatures=config_dict.get("log_model_signatures", False),
+            log_inputs_outputs=config_dict.get("log_inputs_outputs", False),
+            extra_tags=config_dict.get("extra_tags", None),
+        )
+
+
+def patched_inference(func_name, original, self, *args, **kwargs):
+    """
+    A patched implementation of langchain models inference process which enables
+    logging the traces, and other optional artifacts like model, input examples, etc.
+
+    We patch either `invoke` or `__call__` function for different models
+    based on their usage.
+    """
+    # Inject MLflow tracer into the model
+    # TODO: the legacy LangChain callback is removed as its functionality largely
+    #  overlaps with the tracer while increasing the latency due to synchronous
+    #  artifact logging. However, complex text metrics are not logged by the tracer.
+    #  We should recover this functionality either in the tracer or a separate
+    #  callback before merging the LangChain autologging change into the main branch.
+    #  https://github.com/langchain-ai/langchain/blob/38fb1429fe5a955a863fb91b626c2c9f85efb703/libs/community/langchain_community/callbacks/mlflow_callback.py#L62-L81
+    mlflow_tracer = _set_up_langchain_tracer()
+    args, kwargs = _inject_mlflow_callbacks(func_name, [mlflow_tracer], args, kwargs)
+
+    autolog_config = AutoLoggingConfig.init()
+
+    if autolog_config.should_log_optional_artifacts():
+        # This warning is only valid for the legacy LangChain autologging, and the new tracer
+        # should work outside the version range.
+        import langchain
+
+        if not MIN_REQ_VERSION <= Version(langchain.__version__) <= MAX_REQ_VERSION:
+            warnings.warn(
+                "Autologging is known to be compatible with langchain versions between "
+                f"{MIN_REQ_VERSION} and {MAX_REQ_VERSION} and may not succeed with packages "
+                "outside this range."
+            )
+
+        # Logging Traces does not require an active MLflow run, while the other optional artifacts
+        # needs to be logged into an active run, so we need to start a run if it's not already set.
+        run_id = getattr(self, "run_id", _setup_autolog_run(autolog_config))
+
+    with disable_autologging():
+        result = original(self, *args, **kwargs)
+
+    if autolog_config.should_log_optional_artifacts():
+        _log_optional_artifacts(autolog_config, run_id, result, self, func_name, *args, **kwargs)
+
+        # Terminate the run if it is not managed by the user
+        if mlflow.active_run() is None or mlflow.active_run().info.run_id != run_id:
+            mlflow.MlflowClient().set_terminated(run_id)
+
+    return result
+
+
+def _set_up_langchain_tracer():
+    from mlflow.langchain.langchain_tracer import MlflowLangchainTracer
+
+    mlflow_tracer = MlflowLangchainTracer()
+
+    try:
+        mlflow_tracer.flush_tracker()
```

Comment:
This should be called after `result = original(self, *args, **kwargs)`

Reply:
great catch, thanks!